### ISSUE TITLE
More detailed time of day for warnings

### ISSUE THREAD

**Issue by marty2022** [corona-warn-app/cwa-app-android#4613]
## Current Implementation
If you have been near a positive tested person for a longer time you get a warning message with day, month and year when the encounter happened.

## Suggested Enhancement
Beside the date also the time of the day like morning, forenoon, midday, afternoon, evening, night should be displayed so one can better estimate the risk.

## Expected Benefits
As I got the warning I was unable to estimate my own risk:
- Was it in the morning when using public transports while wearing a mask?
- Was it in the forenoon during work while not wearing a mask from the same room, meetings, neighbour rooms or even other companies on the same floor?
- Was it at midday when eating in the cantina while not wearing a mask?
- Was it in the afternoon during work while not wearing a mask from the same room, meetings, neighbour rooms or even other companies on the same floor?
- Was it in the evening when using public transports or shopping while wearing a mask?
- Was it at the night from my neighbours?

So when I would have known a more specific timeframe I would be able to estimate my personal risk much more better and I am also able to specific warn other contact persons.

**Comment by Ein-Tim** [comment]
Hey @marty2022 👋

Thanks for the proposal! 

The Corona-Warn-App is using the Google/Apple Exposure Notification System / Framework. It records and stores the received BLE beacons.
When CWA is checking for risky encounters, it downloads the list of positive keys from the server and hands it over to the Exposure Notification Framework (ENF). This then checks whether a key from the server was also received on your phone. If yes, this means you encountered a later positive tested person. In this case, the ENF hands ofer the relevant Exposure Windows to the CWA. This includes, besides some other things, only the **date of the encounter** but **not** the time. 
So, in conclusion, this proposal is not possible to be implemented without Apple & Google changing their Exposure Notification System.

---

I opened an issue regarding this in the wishlist some time ago: https://github.com/corona-warn-app/cwa-wishlist/issues/266

---

If you want to dive deeper, take a look at the [documentation from Apple on the Exposure Notification Framework](https://developer.apple.com/exposure-notification/) or/and at the [documentation from Google on the Exposure Notification System](https://developers.google.com/android/exposure-notifications/exposure-notifications-api). 

Enjoy Sunday!

**Comment by marty2022** [comment]
What a pity, so changes are very low that this will get changed.
But thank you very much for the detailed explanation!

**Comment by dsarkar** [comment]
Hi @marty2022, yes chances are little... As @Ein-Tim explained already, there are technical limitations that do not allow this. The app uses a functionality (the ENF) from the operating system (Android, or iOS, respectively) in order to assess the risk. 

There is another reasons, why this probably would not be implemented, even if it was technical possible: Data privacy.

On of the main purposes of the app is contact tracing: You as a user should be warned if a possible risk encounter happened. Give gives you the possibility to get advice and most probably you will be sent to get a test (if it was a high risk encounter = red card). For this purpose the time does not add any further useful value.

However, if you had the time, you could maybe identify the person, who shared his positive test result *anonymously*. In this case his privacy would be violated and the acceptance of the app might be decreased.

With the current implementation the app serves amongst others the purpose of contact tracing, i.e. warning user, which then ideally get tested, and on the other side, at the same time it values privacy.

---
Corona-Warn-App Open Source Team

**Comment by cvoinf** [comment]
The fork [corona-contact-tracing-germany](https://codeberg.org/corona-contact-tracing-germany/cwa-android) does allow to get more information. And together with the app  [Corona Warn Companion](https://github.com/mh-/corona-warn-companion-android) you get even more.
![A](https://user-images.githubusercontent.com/11172467/148286549-b3e8f068-f559-416f-93ec-60ac38da689f.png)
![B](https://user-images.githubusercontent.com/11172467/148289818-0daabd09-c551-45bc-9cd9-1b789cf0280a.png)
![C](https://user-images.githubusercontent.com/11172467/148289831-3a3ace00-a587-456f-831c-df7de7f7cb68.png)

**Comment by OlympianRevolution** [comment]
> Hi @marty2022, yes chances are little... As @Ein-Tim explained already, there are technical limitations that do not allow this. The app uses a functionality (the ENF) from the operating system (Android, or iOS, respectively) in order to assess the risk.
> 
> There is another reasons, why this probably would not be implemented, even if it was technical possible: Data privacy.
> 
> On of the main purposes of the app is contact tracing: You as a user should be warned if a possible risk encounter happened. Give gives you the possibility to get advice and most probably you will be sent to get a test (if it was a high risk encounter = red card). For this purpose the time does not add any further useful value.
> 
> However, if you had the time, you could maybe identify the person, who shared his positive test result _anonymously_. In this case his privacy would be violated and the acceptance of the app might be decreased.
> 
> With the current implementation the app serves amongst others the purpose of contact tracing, i.e. warning user, which then ideally get tested, and on the other side, at the same time it values privacy.
> 
> Corona-Warn-App Open Source Team

The problem with this argument is that currently it is easily possible for technically savy users to get the exact timestamp by rooting or installing cctg and then using the companion app to find the exact timesamp.

Why do technically savy people have the possibility (and right?) do deanonymize their contacts via time but not "normal" users?

Either the timestamp should be protected using encryption, hashing or access to key database so that technically savy people also do not have access 
or 
(my preferred solution) the timestamp should be available to everyone.

This seems to be a very frequent feature request.
If Google and Apple have not categorically rejected changes to the ENF one way to convince them might be to make this an opt in/out feature where you only recieve the exact time if you share (allow using) your exact time as well.

**Comment by marty2022** [comment]
@cvoinf Awesome, thank you very much for these links. Too bad that Warn-App-Companion needs root-permissions (or the Corona-Contact-Tracing-App which is not so well tested as they say).

@dsarkar If I only meet one person on a day where I get a positive warning I can also identify this person with the current implementation.
It's always a balance between privacy and benefit. And having 6 timeslots per day instead of 1 would greatly improve the benefit while not decreasing privacy too much.

**Comment by dsarkar** [comment]
Hi @marty2022 

> If I only meet one person on a day where I get a positive warning I can also identify this person with the current implementation.

Sure, in real life scenario, you probably now this person and had been informed by him directly anyway.

> It's always a balance between privacy and benefit. 

That is exactly the point, and current view is that the offered solution is the "right" balance. Of course to find this balance and to agree on common terms is not an easy task. Note that the current ENF by Apple and Google (which limits the functionality to the day of the encounter, and time is unknown by the tracing apps which make use of the OS ENF) is a transnational solution, and many considerations had to go into the definition to achieve acceptance und a working system.

**Comment by Ein-Tim** [comment]
I suggest to move this issue to the wishlist repo.
